src/main/kotlin/biokotlin/cli/MutateProteins.kt

```diff
@@ -0,0 +1,363 @@
+
+        if (rangeToMutate != null) {
+            // get random start index of place to delete
+            // within rangeToMutate
+            val start = Random(4321).nextInt(rangeToMutate.start, rangeToMutate.end - length + 2)
+
+            // delete length bases
+            result.delete(start, start + length)
+        }
+
+        return result.toString()
+
+    }
+
+    /**
+     * This inserts bases (number specified by parameter length) into each sequence.
+     * The bases are continuous. It makes the insertion in or out of
+     * the ranges defined by the bedfile depending on the setting of putMutationsInRanges.
+     */
+    private fun insertionMutation(
+        record: ProteinSeqRecord, ranges: Multimap<String, BedfileRecord>
+    ): String {
+
+        val origSeq = record.seq()
+        val seqLength = origSeq.length
+
+        val contig = record.id
+        val rangesForRecord = ranges.get(contig)
+
+        val mergedRanges = validateAndMergeRanges(seqLength, rangesForRecord)
+
+        val possibleMutationRanges =
+            if (putMutationsInRanges) mergedRanges else inverseRanges(contig, seqLength, mergedRanges)
+
+        // get possible ranges to mutate
+        // Then randomly select one of these ranges
+        val rangeToMutate = possibleMutationRanges
+            .randomOrNull()
+
+        val result = StringBuilder(origSeq)
+
+        if (rangeToMutate != null) {
+            // get random start index of place to insert
+            // within rangeToMutate
+            val start = Random(4321).nextInt(rangeToMutate.start, rangeToMutate.end + 1)
+
+            (0 until length).forEach { _ ->
+                val base = proteinLetters.random()
+                result.insert(start, base)
+            }
+        }
+
+        return result.toString()
+
+    }
+
+    /**
+     * This changes random bases in the sequence to another random value.
+     * It changes upto numMutations bases. It puts the mutations in or out of
+     * the ranges defined by the bedfile depending on the setting of putMutationsInRanges.
+     */
+    private fun pointMutations(
+        record: ProteinSeqRecord,
+        ranges: Multimap<String, BedfileRecord>,
+        mutatedIndicesWriter: BufferedWriter? = null
+    ): String {
+
+        val origSeq = record.seq()
+        val seqLength = origSeq.length
+
+        val contig = record.id
+        val rangesForRecord = ranges.get(contig)
+
+        val mergedRanges = validateAndMergeRanges(seqLength, rangesForRecord)
+
+        val possibleMutateIndices = if (putMutationsInRanges) {
+            origSeq.indices.filter { index -> inRanges(index, mergedRanges) }
+        } else {
+            origSeq.indices.filter { index -> !inRanges(index, mergedRanges) }
+        }.toMutableList()
+
+        val mutatedIndices = mutableSetOf<Int>()
+        var numMutated = 0
+        while (numMutated < numMutations && possibleMutateIndices.isNotEmpty()) {
```

Comment:
Couldn't you use something like this?
`val mutatedIndices = possibleMutateIndices.shuffle().take(numMutations)`

Reply:
That does seem better and concise.  Although kind of don't want to change because I know it works now

Reply:
Testing makes this safer....

There is one difference in the code where the current one can mutate the same bp multiple times potentially.  I am not sure you want that to happen..My version will only do each bp potentially once.